firewood/src/merkle/stream.rs

```diff
@@ -7,399 +7,471 @@ use crate::{
+                    iter_stack.push(IterationNode::Unvisited {
+                        key: matched_key_nibbles.clone().into_boxed_slice(),
+                        node,
+                    });
+                }
+                NodeType::Extension(_) => {
+                    panic!("extension nodes shouldn't exist")
                 }
+            }
 
-                self.key_state = IteratorState::Iterating {
-                    check_child_nibble,
-                    visited_node_path,
+            return Ok(MerkleNodeStreamState::Initialized { iter_stack });
+        };
+        // `nib` is the first nibble after [matched_key_nibbles].
+
+        match &node.inner {
+            NodeType::Branch(branch) => {
+                // The next nibble in `key` is `nib`,
+                // so all children of `node` with a position > `nib`
+                // should be visited since they are after `key`.
+                iter_stack.push(IterationNode::Visited {
+                    key: matched_key_nibbles.clone().into_boxed_slice(),
+                    children_iter: Box::new(
+                        get_children_iter(branch).filter(move |(_, pos)| *pos > nib),
+                    ),
+                });
+
+                // Figure out if the child at `nib` is a prefix of `key`.
+                // (i.e. if we should run this loop body again)
+                #[allow(clippy::indexing_slicing)]
+                let child_addr = match branch.children[nib as usize] {
+                    Some(c) => c,
+                    None => {
+                        // There is no child at `nib`.
+                        // We'll visit `node`'s first child at index > `nib`
+                        // first (if it exists).
+                        return Ok(MerkleNodeStreamState::Initialized { iter_stack });
+                    }
                 };
 
-                self.poll_next(_cx)
-            }
+                matched_key_nibbles.push(nib);
+
+                let child = merkle
+                    .get_node(child_addr)
+                    .map_err(|e| api::Error::InternalError(Box::new(e)))?;
 
-            IteratorState::Iterating {
-                check_child_nibble,
-                visited_node_path,
-            } => {
-                let next = find_next_result(merkle, visited_node_path, check_child_nibble)
-                    .map_err(|e| api::Error::InternalError(Box::new(e)))
-                    .transpose();
+                let partial_key = match child.inner() {
+                    NodeType::Branch(branch) => &branch.path,
+                    NodeType::Leaf(leaf) => &leaf.path,
+                    NodeType::Extension(_) => {
+                        panic!("extension nodes shouldn't exist")
+                    }
+                };
 
-                Poll::Ready(next)
+                let (comparison, new_unmatched_key_nibbles) =
+                    is_prefix(partial_key.iter(), unmatched_key_nibbles);
```

Comment:
nit: 
`is_prefix` implies a `bool` as a return type. I say this as someone who frequently gives functions terrible names. What about `compare_as_prefix` or something along those lines?

Reply:
renamed `compare_partial_path`